feat: emit diagnostics_channel events upon routing request

tlhunter opened the pull request:
This adds support for emitting [diagnostics channel](https://nodejs.org/api/diagnostics_channel.html#class-tracingchannel) events for a request. It's an extension of #5105. 

The `dc-polyfill` package provides a polyfill of the `diagnostics_channel` module for older versions of Node.js. This is why the existing diagnostics channel call has been removed from a conditional in `fastify.js`. Note that `dc-polyfill` falls back to the existing diagnostics channel if present. That's why the tests are requiring diagnostics channel directly instead of the polyfill to show that it works as expected.

Here's some things I'm wondering about:

- [x] extracting payload from `reply.send()`
- [x] rebased on main
- [x] tests
- [x] docs

#### Benchmark results

**main**

```
┌─────────┬──────┬──────┬───────┬───────┬────────┬─────────┬────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg    │ Stdev   │ Max    │
├─────────┼──────┼──────┼───────┼───────┼────────┼─────────┼────────┤
│ Latency │ 4 ms │ 9 ms │ 13 ms │ 13 ms │ 8.3 ms │ 4.19 ms │ 365 ms │
└─────────┴──────┴──────┴───────┴───────┴────────┴─────────┴────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬────────────┬──────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg        │ Stdev    │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼────────────┼──────────┼─────────┤
│ Req/Sec   │ 104,639 │ 104,639 │ 114,495 │ 116,735 │ 113,796.27 │ 2,451.61 │ 104,582 │
├───────────┼─────────┼─────────┼─────────┼─────────┼────────────┼──────────┼─────────┤
│ Bytes/Sec │ 19.7 MB │ 19.7 MB │ 21.5 MB │ 22 MB   │ 21.4 MB    │ 461 kB   │ 19.7 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴────────────┴──────────┴─────────┘
```

**this change**

```
┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max    │
├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼────────┤
│ Latency │ 4 ms │ 9 ms │ 13 ms │ 13 ms │ 8.21 ms │ 4.23 ms │ 364 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬───────────┬──────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg       │ Stdev    │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼───────────┼──────────┼─────────┤
│ Req/Sec   │ 107,455 │ 107,455 │ 115,455 │ 117,503 │ 114,822.4 │ 2,261.46 │ 107,414 │
├───────────┼─────────┼─────────┼─────────┼─────────┼───────────┼──────────┼─────────┤
│ Bytes/Sec │ 20.2 MB │ 20.2 MB │ 21.7 MB │ 22.1 MB │ 21.6 MB   │ 426 kB   │ 20.2 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴───────────┴──────────┴─────────┘
```

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

tlhunter commented:
/cc @Qard

Qard commented on lib/handleRequest.js:
```diff
@@ -125,29 +128,42 @@ function validationCompleted (request, reply, validationErr) {
     reply[kReplyIsError] = true
     reply.send(err)
+    store.error = err
+    channels.end.publish(store)
```
I assume this should be `error`?

```suggestion
    channels.error.publish(store)
```

Qard commented on lib/handleRequest.js:
```diff
@@ -125,29 +128,42 @@ function validationCompleted (request, reply, validationErr) {
+
     reply[kReplyIsError] = true
     reply.send(err)
+    channels.end.publish(store)
```
Rather than putting the `end` in all the branches you could use try/finally. That also ensures the event still fires if something throws.

Qard commented on lib/wrapThenable.js:
```diff
@@ -28,6 +33,8 @@ function wrapThenable (thenable, reply) {
       }
     }
+
+    if (channels) channels.asyncEnd.publish(store)
```
Should do try/finally here too.

Qard reviewed (changes requested):
Left a few suggestions. Basically try/finally to more reliably ensure end events happen. Should probably also just put a catch in there too to be sure any unexpected errors get reported too.

mcollina commented on fastify.js:
```diff
@@ -526,15 +526,10 @@ function fastify (options) {
-  } catch (e) {
-    // This only happens if `diagnostics_channel` isn't available, i.e. earlier
-    // versions of Node.js. In that event, we don't care, so ignore the error.
+  const dc = require('dc-polyfill')
```
move this to the top of the file

mcollina commented on lib/wrapThenable.js:
```diff
@@ -5,12 +5,17 @@ const {
+    if (channels) {
+      store.result = payload
+      channels.asyncStart.publish(store)
+    }
```
I'm not 100% convinced of the meaning of asyncStart. Can you clarify why you put it here?

Qard replied:
With callbacks there's asyncStart and asyncEnd scoping the callback. Most tracers want to create a span up to the start of the callback, but some consider the callback to be part of the task and so want to capture up to the end. In the case of promises there's no distinct "end" condition other than GC so it generally emits the two together for promises. I think the logic here was that it's surrounding the _internal_ continuation of the thenable rather than just have the two events right beside each other without scoping _anything_ due to that limitation of not knowing the _actual_ end. Does that makes sense here?

mcollina reviewed (commented):
given dc-polyfill, I think we can land this on main if it does not introduce overhead when disabled.

mcollina commented:
> Is there a way to get the payload sent via reply.send()?

inside `reply.send()` or when executing the onSend hook.

Uzlopak commented:
What about the diagnostics_channel plugin?

jsumners commented on docs/Reference/Hooks.md:
```diff
@@ -827,18 +827,18 @@ consider creating a custom [Plugin](./Plugins.md) instead.
+> versions of Node.js are supported by Fastify where `diagnostics_channel` is
+> unavailable, or with an incomplete feature set, the hook uses the
+> [dc-polyfill](https://www.npmjs.com/package/dc-polyfill) package to provide a
+> polyfill. Feel free to either use this package or `diagnostics_channel` itself
+> in your application.
```
What benefit is the final sentence providing to the Fastify docs?

```suggestion
> polyfill.
```

tlhunter replied:
An application would benefit if it needs to run on a wide range of Node.js versions. But, since app owners have more control over Node.js, the point is moot. I'll apply your suggested removal.

jsumners commented on docs/Reference/Hooks.md:
```diff
@@ -865,3 +865,37 @@ channel.subscribe(function ({ fastify }) {
+
+Five other events are published on a per-request basis following the
+[Tracing Channel](https://nodejs.org/api/diagnostics_channel.html#class-tracingchannel)
+nomenclature. Here's a list of the channel names and the event they receive:
```
```suggestion
nomenclature. The list of the channel names and the event they receive is:
```

jsumners commented on docs/Reference/Hooks.md:
```diff
@@ -865,3 +865,37 @@ channel.subscribe(function ({ fastify }) {
+returned / resolved response. Finally, the `:error` event contains an `error`
+property associated with the request's failure.
+
+These events can be received by doing the following:
```
```suggestion
These events can be received like so:
```

Uzlopak commented:
Clarifying my last comment. How does this PR relate to

https://github.com/fastify/fastify-diagnostics-channel

?

Seems like this PR makes the plugin partially redundant?

Qard commented:
Definitely overlaps with the plugin, though this is more limited to just producing a trace around a request whereas the plugin gives a bit more control. This might benefit from a few more events to cover the other use cases provided by the plugin.

The main benefit I see for this existing rather than just the plugin is that if it lives directly in the library it enables automatic instrumentation tracers to listen without producing side effects while adding the plugin could have user-visible behaviour. There's also probably performance advantages to going direct rather than going through the plugin.

jsumners commented:
I think instrumentation hooks such as these should be in core. They provide a way for other instrumentations to do things via a specific plugin or wrapping package. If we were trying to add such hooks via some third party library, like dtrace, then my opinion would be different. But the API used in this PR is meant to be a core API.

tlhunter commented:
> What about the diagnostics_channel plugin?

@Uzlopak I think you're asking about the removal of the `diagnostics_channel` npm package?

If so, the `dc-polyfill` package makes the `diagnostics_channel` package obsolete. The way `diagnostics_channel` worked is that it was a snapshot of the internal `diagnostics_channel` module that was made into an npm package of the same name. On ancient Node.js versions where the internal `diagnostics_channel` module was missing a call to `require('diagnostics_channel')` would pickup the npm package and that same call on newer versions of Node.js simply retrieves the internal module. However there were shortcomings with the `diagnostics_channel` package, like if multiple versions of the `diagnostics_channel` package existed in the a `node_modules/` dependency tree then they wouldn't cross communicate. Also the `diagnostics_channel` package hasn't been maintained and is missing newer features that this PR depends on. `dc-polyfill` alleviates those issues and many more.

If anyone is curious please checkout the [dc-polyfill](https://github.com/DataDog/dc-polyfill) docs for more info. Ultimately `dc-polyfill` is a drop-in replacement for both the `diagnostics_channel` npm package and internal module

Qard commented:
> I think instrumentation hooks such as these should be in core. They provide a way for other instrumentations to do things via a specific plugin or wrapping package. If we were trying to add such hooks via some third party library, like dtrace, then my opinion would be different. But the API used in this PR is meant to be a core API. 

It _is_ a core API. The dc-polyfill API just enables supporting the newer features beyond the limited backporting window of Node.js core. With dc-polyfill we had it tested and working all the way back to Node.js 12. It uses the core API where available though, so it should always be a drop-in replacement for using the core API as if using the _current_ version of Node.js.

As for the plugin versus embedding in the library, I much favour embedding in the library as the diagnostics_channel API (and dc-polyfill on top of it) are aimed at very high performance and as close to zero cost as possible when nothing is listening. Abstracting that with a plugin in the middle adds a bunch of unnecessary overhead as fastify then loses the context of what is ignorable as the plugin barrier isn't designed to forward that information.

I created diagnostics_channel and it is quite specifically my intent that it be used directly in libraries in this way as it gets us as close to the source as possible feed of diagnostics data, eliminating a ton of instrumentation cost in the process and making gathering instrumentation insights much, much easier in the process. The existing model of monkey-patching everything is highly unreliable and very unapproachable to an average developer to be able to use as a way to gain insight about their applications. I intended for diagnostics_channel to encourage library developers to start thinking about exposing diagnostics data directly and thinking about the performance profile of their code more.

tlhunter commented:
> inside `reply.send()` or when executing the onSend hook.

@mcollina I'm now grabbing the payload within `Reply` instances and setting it on itself using a new symbol: https://github.com/fastify/fastify/pull/5252/commits/92f77e4b1d183a873bc750ded6710514883b9052

That said I'm not sure if there are expectations on how long `Reply` instances stick around and if there may be memory concerns. Let me know if this approach isn't ideal and I'll try something else.

Uzlopak commented:
I wonder if we should integrate then the whole diagnostics-channel stuff from the plugin into core.

climba03003 commented on lib/reply.js:
```diff
@@ -66,6 +67,7 @@ function Reply (res, request, log) {
   this[kReplyTrailers] = null
   this[kReplyHasStatusCode] = false
   this[kReplyStartTime] = undefined
+  this[kReplyPayload] = undefined
```
Only concern is when the payload is `stream`.
How can the subscriber safely consume the `stream` for logging?
The same problem exist when I implement the `Trailer` feature and it is not resolved until now.

mcollina replied:
I can see so many ways this can end badly long term. Can you avoid storing the payload inside reply?

tlhunter replied:
The latest commit no longer assigns the payload value to the `Reply` instance: 325ba542cd19759cf18eb69b6b11c116e462f70e

mcollina commented:
I think you can target main for this.

mcollina commented:
>  I wonder if we should integrate then the whole diagnostics-channel stuff from the plugin into core. 

Agreed

tlhunter commented on lib/handleRequest.js:
```diff
@@ -158,7 +163,6 @@ function preHandlerCallback (err, request, reply) {
       if (result !== null && typeof result.then === 'function') {
         wrapThenable(result, reply, channels, store)
       } else {
-        store.result = result
```
This line becomes redundant as the `kReplyOnSend` callback sets the value.

tlhunter commented:
> I think you can target main for this.

This PR is now based on main. Ended up squashing due to conflicts.

mcollina commented on lib/wrapThenable.js:
```diff
@@ -5,28 +5,37 @@ const {
 } = require('./symbols')
 
-function wrapThenable (thenable, reply) {
+function wrapThenable (thenable, reply, channels, store) {
```
I don't think this function should be modified, I would recommend to keep the change between `preHandlerCallback` and `reply.send()`.
Specifically, I would emit `asyncStart` and `asyncEnd` within `reply.send()`, as there are multiple ways to call that function (synchronous and asynchronous)

mcollina reviewed (changes requested):
Good work! I think a few edge cases are missing:

1. [ ] `reply.callNotFound()` // this will execute the not found handler
2. [ ] Fastify error handlers (nested)
3. [ ] routes with `setImmediate(() => reply.send('hello world))`
4. [ ] async routes with `setImmediate(() => reply.send('hello world)); return reply`

The reason why I'm asking is that it will _definitely_ pass through `reply.send()` multiple times,

Uzlopak commented:
@mcollina 

Should this PR also integrate the plugins behaviour? Or should this be a follow up?

mcollina commented:
> @mcollina 
> 
> 
> 
> Should this PR also integrate the plugins behaviour? Or should this be a follow up?

Can you clarify?

tlhunter commented:
@mcollina I think @Uzlopak is asking if I should copy all of the code from the existing `fastify-diagnostics-channel` package and incorporate it into my PR. I'm of the opinion that it shouldn't be included in this PR. While that package is semi-related in scope, it uses a different pattern, and this PR is just attempting to implement the TracingChannel "spec". Plus I'm going to write some blog posts that link to this PR to share with the community and I'd like to keep the PR as succinct and isolated as possible to avoid confusing the audience.

mcollina commented:
A follow up would be good, this would be massive already after the fixes I mentioned.

tlhunter commented:
@mcollina Just so I know if I'm on the right track, do you think I need to refactor a bunch of Fastify to achieve those changes?

tlhunter commented:
@Qard and I paired on this PR today. He envisions two levels of tracing channel in web frameworks. The outer level is that of routing, and the inner level is that of request handling. Previously I was taking a sort of hybrid approach. This PR has been refactored to implement tracing at the routing level. For that reason code has been pulled out of `reply.js` and moved into `wrapThenable.js`.

The response payload isn't necessarily required for APM tools, and with the way TracingChannel is designed it's a bit kludgy to extract using `res.send()` anyway. While it's easy to extract in some scenarios (returns/resolves) I've modified the PR to not include `.result` in the emitted store objects to reduce confusion.

Adding tracing at the request level is something that can be added in the future. The two will play nicely with each other at that point. Request tracing would require the changes to `res.send()` that I was previously working on. Request level tracing is mostly covered by http tracing which is why it's a lower priority.

Since this is the routing layer tracing I also updated the stores to include information about the routing decision made by Fastify. Specifically the underlying route pattern and HTTP method.

Tracing plugins/middleware would be valuable too but we're currently considering them to be out of scope.

Qard commented on lib/handleRequest.js:
```diff
@@ -125,29 +128,54 @@ function validationCompleted (request, reply, validationErr) {
+          store.async = true
+          wrapThenable(result, reply, store)
+        } else {
+          // start and end should always happen before async start/end
+          reply.send(result) // maybe just universally use reply.send and never check response since that's handled here
```
Did you mean to leave the comments?

tlhunter replied:
Whoops, removed.

Qard commented on lib/reply.js:
```diff
@@ -131,6 +131,7 @@ Reply.prototype.hijack = function () {
 }
 
 Reply.prototype.send = function (payload) {
+  // console.trace('Reply#send()', payload)
```
I assume you didn't mean to leave this?

tlhunter replied:
Whoops, removed.

mcollina commented on lib/wrapThenable.js:
```diff
@@ -5,42 +5,67 @@ const {
+    try {
+      if (reply[kReplyHijacked] === true) {
+        return
+      }
```
this should happen before asyncStart, because otherwise there will be no asyncEnd.

Qard replied:
That's what the try/finally is for. I'm not too familiar with exactly what the purpose is of the `kReplyHijacked` thing there, but I'm going on the assumption that even when the early exit happens we'd still want to see the asyncStart/asyncEnd?

mcollina replied:
you won't get an asyncEnd if hijacked.

Qard replied:
Why? Wouldn't the try/finally ensure the finally _must_ happen? 🤔

mcollina reviewed (commented):
I think this is solid, one tiny bit left.

mcollina commented:
I think the nomenclature and names do not match what we use in Fastify. I would identify that specific event as “request.handler”, because “rout ing” happens in a completely different place.

Uzlopak commented:
The tests are failing and seem to be a valid report, because handleRequest.js was modified by this PR.

tlhunter commented:
> I think the nomenclature and names do not match what we use in Fastify. I would identify that specific event as “request.handler”, because “rout ing” happens in a completely different place.

@mcollina I've renamed `fastify.route` to `fastify.request.handler`. Is that what you had in mind?

mcollina commented:
Good work, I think we are almost there.

I would like to have tests for two cases:

1. "nested error handlers" (plugins with an error handler that throws.. this is catched by a parent error handler)
2. route that calls `callNotFound()`

Those will generate more events, and we need to verify the behavior.

Also, could you move all the dc tests into its own folder?

tlhunter commented:
I'm currently trying to work out some issues with the 404 handler where the start event fires twice.

simon-id commented:
FYI @mcollina I'm planning on opening a follow up PR after this one is merged to introduce an "aborting" mechanism to the DC events. I'd like to have your web framework maintainer feedback before making it a "standard". I don't want to block this PR from landing meanwhile :+1:

metcoder95 commented:
>I'm planning on opening a follow up PR after this one is merged to introduce an "aborting" mechanism to the DC events. 

Interested, what do you refer to with `aborting`?

Qard commented:
The use case for aborting is security products identifying attacks and aborting the requests early.

mcollina commented:
Sounds good.

simon-id commented:
>Interested, what do you refer to with aborting?

So sorry I wrote a lengthy answer and forgot to send it :facepalm: 

A code snippet is worth a thousand words:
```javascript
function requestHandler (req, res) {
  const abortController = new AbortController()

  requestStartChannel.publish({ req, res, abortController })

  if (abortController.signal.aborted) {
    // do not pass request to user handler
  } else {
    // call user handler
  }
}
```
Basically our thoughts is that DC events should be more than just read-only. So make sure passed data is mutable, and provide a way to "cancel" the current operation.
For the record I've opened [a PR in node core](https://github.com/nodejs/node/pull/38917) to integrate this function in the API directly, but it didn't get enough traction. So I'd like to reintroduce it as a "soft" standard instead, and use Fastify as a pilot experiment.

(Disclaimer, both [Qard](https://github.com/Qard) and [tlhunter](https://github.com/tlhunter) are my coworkers)

metcoder95 commented:
Nice, thanks for the snippet and the context, SGTM 👍

mcollina commented on lib/handleRequest.js:
```diff
@@ -125,29 +128,53 @@ function validationCompleted (request, reply, validationErr) {
+      channels.end.publish(store)
     }
-  }
+  })
```
We were _very_ careful of not creating an added closure in this function. Closures creates overhead and this is _very_ hot. This should removed.

simon-id replied:
`runStores()` accepts arguments to be passed to the wrapped function. So we could use at L143 something like:
```javascript
channels.start.runStores(store, _preHandlerCallback, this, err, request, reply)
```
but looking at the implementation in `dc-polyfill`, it just creates a closure internally: https://github.com/DataDog/dc-polyfill/blob/3a018599e2745ea3373ed151ab525d80f286e669/patch-channel-store-methods.js#L36
So I'm guessing that defeats the point.

tlhunter replied:
@mcollina I removed the closure in Fastify but Simon is right in that one still exists within diagnostics channel.

simon-id replied:
@tlhunter Is it even possible to get rid of all closure in a case like this? Node's ALS has `enterWith()` but I'm not sure it's faster than a closure and it's not been kept by the tc39 spec anyway :/

tlhunter replied:
@simon-id one thing I could do is check all five channels for listeners. If none are present then run `preHandlerCallbackInner()` without the store argument and without first calling `runStores()`. This prevents the closure in most Fastify use-cases.

tlhunter replied:
Pretty sure I fixed this issue

mcollina reviewed (changes requested):
lgtm

tlhunter commented on lib/handleRequest.js:
```diff
@@ -125,28 +129,64 @@ function validationCompleted (request, reply, validationErr) {
   }
 
-  let result
+  if (context[kFourOhFourContext] === null) {
```
@mcollina I was able to fix the issue with 404 routes executing twice with this check. But I'm wondering if the solution is now too-specific to handling 404 errors and if other situations can trigger the same problem?

mcollina replied:
Possibly the error path, but I think you handled that already.

mcollina commented on lib/handleRequest.js:
```diff
@@ -125,28 +129,64 @@ function validationCompleted (request, reply, validationErr) {
+    route: {
+      url: context.config.url,
+      method: context.config.method
+    }
```
Can we avoid allocating those two objects if there are no diagnostics channels?

tlhunter replied:
The creation of the store object is now wrapped in a check to see if there are any listeners on any of the channels.

mcollina reviewed (commented):
Good job, just one nit and this should be good to go

tlhunter commented:
@mcollina I modified the code to check for the presence of listeners on any of the DC channels. If there aren't any listeners then the code skips the store object creation paths and also skips the `runStores()` function call which adds a closure. Here's an updated benchmark run:

### main

```
┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼───────┤
│ Latency │ 4 ms │ 9 ms │ 13 ms │ 13 ms │ 8.08 ms │ 2.84 ms │ 62 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬────────────┬──────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg        │ Stdev    │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼────────────┼──────────┼─────────┤
│ Req/Sec   │ 105,535 │ 105,535 │ 117,695 │ 120,639 │ 116,494.94 │ 3,278.49 │ 105,482 │
├───────────┼─────────┼─────────┼─────────┼─────────┼────────────┼──────────┼─────────┤
│ Bytes/Sec │ 19.8 MB │ 19.8 MB │ 22.1 MB │ 22.7 MB │ 21.9 MB    │ 619 kB   │ 19.8 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴────────────┴──────────┴─────────┘
```

### tlhunter/diagnostics-channel

```
┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼───────┤
│ Latency │ 4 ms │ 9 ms │ 13 ms │ 13 ms │ 8.01 ms │ 2.78 ms │ 56 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬────────────┬──────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg        │ Stdev    │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼────────────┼──────────┼─────────┤
│ Req/Sec   │ 103,295 │ 103,295 │ 118,847 │ 121,023 │ 117,486.94 │ 3,489.57 │ 103,245 │
├───────────┼─────────┼─────────┼─────────┼─────────┼────────────┼──────────┼─────────┤
│ Bytes/Sec │ 19.4 MB │ 19.4 MB │ 22.3 MB │ 22.8 MB │ 22.1 MB    │ 656 kB   │ 19.4 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴────────────┴──────────┴─────────┘
```

Qard commented on lib/handleRequest.js:
```diff
@@ -125,30 +129,73 @@ function validationCompleted (request, reply, validationErr) {
+
+    if (result !== undefined) {
+      if (result !== null && typeof result.then === 'function') {
+        if (store) store.async = true
```
Might as well put this inside `wrapThenable` if you're passing in `store` already anyway. 🤔

Qard commented on lib/wrapThenable.js:
```diff
@@ -5,42 +5,67 @@ const {
+
+      reply[kReplyIsError] = true
+
+      // try-catch allow to re-throw error in error handler for async handler
+      try {
+        reply.send(err)
+      } catch (err) {
+        reply.send(err)
+      }
```
Could probably just raise this try/catch into a merge with the outer try/finally.

Qard reviewed (approved):
LGTM, just some minor nits, but feel free to ignore them. 😄

mcollina reviewed (approved):
lgtm

mcollina commented:
Can you run a final bench?

tlhunter commented:
@mcollina I ran it a couple times and the results are always about the same:

```
$ clear ; gc tlhunter/diagnostics-channel && npm run benchmark && gc main && npm run benchmark
Switched to branch 'tlhunter/diagnostics-channel'
Your branch is up to date with 'tlhunter/tlhunter/diagnostics-channel'.

> fastify@4.26.2 benchmark
> concurrently -k -s first "node ./examples/benchmark/simple.js" "autocannon -c 100 -d 30 -p 10 localhost:3000/"

[1] Running 30s test @ http://localhost:3000/
[1] 100 connections with 10 pipelining factor
[1]
[1]
[1] ┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬───────┐
[1] │ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max   │
[1] ├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼───────┤
[1] │ Latency │ 4 ms │ 9 ms │ 12 ms │ 13 ms │ 7.77 ms │ 2.73 ms │ 55 ms │
[1] └─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴───────┘
[1] ┌───────────┬─────────┬─────────┬─────────┬─────────┬────────────┬──────────┬─────────┐
[1] │ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg        │ Stdev    │ Min     │
[1] ├───────────┼─────────┼─────────┼─────────┼─────────┼────────────┼──────────┼─────────┤
[1] │ Req/Sec   │ 104,511 │ 104,511 │ 122,175 │ 124,735 │ 120,834.14 │ 4,364.21 │ 104,463 │
[1] ├───────────┼─────────┼─────────┼─────────┼─────────┼────────────┼──────────┼─────────┤
[1] │ Bytes/Sec │ 19.6 MB │ 19.6 MB │ 23 MB   │ 23.5 MB │ 22.7 MB    │ 822 kB   │ 19.6 MB │
[1] └───────────┴─────────┴─────────┴─────────┴─────────┴────────────┴──────────┴─────────┘
[1]
[1] Req/Bytes counts sampled once per second.
[1] # of samples: 30
[1]
[1] 3626k requests in 30.02s, 681 MB read
[1] autocannon -c 100 -d 30 -p 10 localhost:3000/ exited with code 0
--> Sending SIGTERM to other processes..
[0] node ./examples/benchmark/simple.js exited with code SIGTERM
Switched to branch 'main'
Your branch is up to date with 'origin/main'.

> fastify@4.26.2 benchmark
> concurrently -k -s first "node ./examples/benchmark/simple.js" "autocannon -c 100 -d 30 -p 10 localhost:3000/"

[1] Running 30s test @ http://localhost:3000/
[1] 100 connections with 10 pipelining factor
[1]
[1]
[1] ┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬───────┐
[1] │ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max   │
[1] ├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼───────┤
[1] │ Latency │ 4 ms │ 9 ms │ 12 ms │ 13 ms │ 7.95 ms │ 2.77 ms │ 61 ms │
[1] └─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴───────┘
[1] ┌───────────┬─────────┬─────────┬─────────┬─────────┬───────────┬──────────┬─────────┐
[1] │ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg       │ Stdev    │ Min     │
[1] ├───────────┼─────────┼─────────┼─────────┼─────────┼───────────┼──────────┼─────────┤
[1] │ Req/Sec   │ 106,495 │ 106,495 │ 118,591 │ 121,727 │ 118,182.4 │ 3,054.91 │ 106,448 │
[1] ├───────────┼─────────┼─────────┼─────────┼─────────┼───────────┼──────────┼─────────┤
[1] │ Bytes/Sec │ 20 MB   │ 20 MB   │ 22.3 MB │ 22.9 MB │ 22.2 MB   │ 574 kB   │ 20 MB   │
[1] └───────────┴─────────┴─────────┴─────────┴─────────┴───────────┴──────────┴─────────┘
[1]
[1] Req/Bytes counts sampled once per second.
[1] # of samples: 30
[1]
[1] 3546k requests in 30.02s, 667 MB read
[1] autocannon -c 100 -d 30 -p 10 localhost:3000/ exited with code 0
--> Sending SIGTERM to other processes..
[0] node ./examples/benchmark/simple.js exited with code SIGTERM
```

mcollina commented:
@jsumners @Eomm ptal

jsumners commented on lib/handleRequest.js:
```diff
@@ -125,30 +129,72 @@ function validationCompleted (request, reply, validationErr) {
 }
 
+function tcHasSubscribers () {
+  return channels.start.hasSubscribers ||
+    channels.end.hasSubscribers ||
+    channels.asyncStart.hasSubscribers ||
+    channels.asyncEnd.hasSubscribers ||
+    channels.error.hasSubscribers
```
@Qard do you agree that there should be a `tracingChannel.hasSubscribers() === true` method in core so that implementations don't have to maintain a function like this one?

Qard replied:
Yes, just landed it the other day with the change to enable early exits. It's not in dc-polyfill yet though.

jsumners reviewed (approved):
Looks good to me. Just a note for upstream API design.

Uzlopak commented:
What i really dont understand is that according to @Qard the publish function is a noop when no subscriber exist. So then why do we even have to check for hasSubscribers?

Qard commented:
The hasSubscribers helper is for the publisher to be able to skip message construction if they are creating a new object for the message. If they are just passing an existing object it's not necessary, but here we _are_ creating a new object.

tlhunter commented:
> What i really dont understand is that according to @Qard the publish function is a noop when no subscriber exist. So then why do we even have to check for hasSubscribers?

It was to address Matteo's object creation concern: https://github.com/fastify/fastify/pull/5252#discussion_r1505817108

tlhunter commented:
FYI, I'll probably have @Qard address any follow up concerns with this PR.

jsumners commented:
Can we get the helper function added to `dc_polyfill` and then utilized in this PR before merging?

Qard commented:
The PR for it is here: https://github.com/DataDog/dc-polyfill/pull/10. Needs some work yet as tests were failing on some versions though. Haven't yet identified why as I've been busy with other things for a bit. 😅

mcollina commented:
In v5, our minimum supported version will be _latest_ v18.

tlhunter commented:
I'm back from parental leave and will be working on landing this PR again. It looks like @Qard was working on landing a change in dc-polyfill but hit a roadblock. I'll look at his PR and see if I can help out there.

tlhunter commented:
I'll focus on the new 404 error first.

tlhunter commented:
I had a few questions for @Qard but he just went on vacation so I'll try to figure things out without him. Here's a paper trail of what's going on:

1. @Qard [landed a commit](https://github.com/nodejs/node/pull/51915) in Node.js versions 22.0.0 and 20.13.0 which does two things:
    - adds the `TracingChannel#hasSubscribers` helper method, [as requested above](https://github.com/fastify/fastify/pull/5252/#issuecomment-2001993669)
    - adds some early-exit functionality which is a little more difficult to implement [as eluded by @Qard above](https://github.com/fastify/fastify/pull/5252/#issuecomment-2008157521) 
2. I now [have a PR open](https://github.com/DataDog/dc-polyfill/pull/11) to add just the `hasSubscribers` method to dc-polyfill while we figure out the early-exit part
    - I'll run it by my team since @Qard is out

If I can merge the dc-polyfill PR today then I can update this Fastify PR and AFAICT it'll be ready to land.

tlhunter commented:
The backported functionality has landed in dc-polyfill as of v0.1.5 and this PR has been updated to make use of the new helper method.

tlhunter commented:
@mcollina do you think this will be able to land in the v5 release?

mcollina commented:
can you rebase and make CI green? I've restarted it

mcollina commented:
This can land I think

jsumners reviewed (approved):
Looks good to me.

jsumners commented:
https://github.com/fastify/fastify/actions/runs/9114429260/job/25058583913?pr=5252#step:5:17959

There is a test timing out.

tlhunter commented:
I'm able to reproduce this locally but only if I run via c8:

```
PASS: ./node_modules/.bin/tap
FAIL: ./node_modules/.bin/c8 ./node_modules/.bin/tap 
PASS: node ./test/wrapThenable.test.js
HANG: ./node_modules/.bin/c8 node ./test/wrapThenable.test.js
PASS: ./node_modules/.bin/c8 node ./test/listen.3.test.js # random other test
```

Notably the individual tests pass but something about c8 hangs indefinitely. I'm still looking into it.

jsumners commented:
Is it only happening on 18? Does it pass on 20+?

tlhunter commented:
Looks like it also happens on the oldest and newest 20, and not the last 16.

jsumners commented:
Drat! If it were only on 18, we had an easy solution.

mcollina commented:
Hopefully the problem with the v18 FinalizationRegistry timeouts will be fixed with the upcoming release, the fix has _finally_ been backported.

tlhunter commented:
I rebased and kicked off the CI jobs again. It appears that fewer are failing:

![2024-05-30_170408](https://github.com/fastify/fastify/assets/551402/80bf033a-23f8-48ca-b700-81179eabb8db)

I did some more testing locally:

```
v20.11.1 HANG
v20.12.0 OK 2024-03-26

v18.19.1 HANG
v18.20.0 HANG
v18.20.1 HANG
v18.20.2 HANG
v18.20.3 OK 2024-05-21
```

Currently it looks like [CI is running the final failing tests against v18.20.2](https://github.com/fastify/fastify/actions/runs/9184228339/job/25256278726?pr=5252). Hopefully it'll disappear from the cache soon and the PR will turn green?

![2024-05-30_171202](https://github.com/fastify/fastify/assets/551402/51babec2-4ab1-433b-a437-6214ed084bc5)

tlhunter commented:
@jsumners @mcollina either someone cleared the cache or I checked an older job... Looks like everything is green now!

mcollina reviewed (approved):
lgtm
